web/packages/teleport/src/Navigation/RecentHistory.tsx

```diff
@@ -0,0 +1,176 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+import React, { useState } from 'react';
+import styled from 'styled-components';
+import { matchPath } from 'react-router';
+import { NavLink } from 'react-router-dom';
+
+import { Flex, Text, Box, H4 } from 'design';
+import { Cross } from 'design/Icon';
+
+import { TeleportFeature } from 'teleport/types';
+import { useFeatures } from 'teleport/FeaturesContext';
+
+import { getSubsectionStyles } from './SideNavigation/Section';
+import { SidenavCategory } from './SideNavigation/categories';
+
+/**
+ * RecentHistoryItem is an item in the recent history list, this is the data format stored in the recent history kept in localStorage.
+ */
+export type RecentHistoryItem = {
+  category?: SidenavCategory;
+  title: string;
+  route: string;
+  exact?: boolean;
+};
+
+function getIconForRoute(
+  features: TeleportFeature[],
+  route: string
+): (props) => JSX.Element {
+  const feature = features.find(feature =>
+    matchPath(route, {
+      path: feature?.route?.path,
+      exact: false,
+    })
+  );
+
+  return feature?.navigationItem?.icon || feature?.topMenuItem?.icon;
+}
+
+export function RecentHistory({
+  recentHistoryItems,
+  onRemoveItem,
+}: {
+  recentHistoryItems: RecentHistoryItem[];
+  onRemoveItem: (route: string) => void;
+}) {
+  const features = useFeatures();
+
+  return (
+    <Flex flexDirection="column" mt={3} width="100%" gap={1}>
+      {!!recentHistoryItems.length && (
+        <>
+          <H4 px={3} color="text.muted" style={{ textTransform: 'none' }}>
+            Recent Pages
+          </H4>
+          {recentHistoryItems.map((recentHistoryItem, index) => {
+            const Icon = getIconForRoute(features, recentHistoryItem.route);
+
+            return (
+              <RecentHistoryItem
+                key={index}
+                recentHistoryItem={recentHistoryItem}
+                Icon={Icon}
+                onRemove={() => onRemoveItem(recentHistoryItem.route)}
+              />
+            );
+          })}
+        </>
+      )}
+    </Flex>
+  );
+}
+
+function RecentHistoryItem({
+  recentHistoryItem,
+  Icon,
+  onRemove,
+}: {
+  recentHistoryItem: RecentHistoryItem;
+  Icon: (props) => JSX.Element;
+  onRemove: () => void;
+}) {
+  const [hovered, setHovered] = useState(false);
+
+  return (
+    <RecentHistoryItemWrapper
+      as={NavLink}
+      to={recentHistoryItem.route}
+      onMouseEnter={() => setHovered(true)}
+      onMouseLeave={() => setHovered(false)}
+    >
+      <Flex width="100%" gap={2} alignItems="start">
+        <Flex height="24px" alignItems="center" justifyContent="center">
+          <Icon size={20} color="text.slightlyMuted" />
```

Comment:
this is the one that could be undefined. If every route had a navigation item or topmenuicon i'd say just make one of them NOT optional in the type, but i don't think thats the case. I think something as simple as `Icon && <Icon ... />` is probably fine here and we'll just catch the ones that dont have icons visually when we dev it